### PR TITLE
Added UDL support. Changed string to vector function using regex.

### PR DESCRIPTION
**Pull request by SagnikDey92**
Please review. If more tests for UDL required, let me know and I'll update the PR.

### CODE REVIEW

File: include/real/real_explicit.hpp [hunk 1]
```diff
@@ -221,102 +222,82 @@ namespace boost {
              */
             real_explicit(const real_explicit& other)  = default;
 
-            /**
```
Comment (Laouen) [inline include/real/real_explicit.hpp]: why did you remove the function documentation?

File: include/real/real_explicit.hpp [hunk 2]
```diff
@@ -221,102 +222,82 @@ namespace boost {
...
+                    else
+                        decimal_part = number.substr(dot_pos+1);
+                }
+                if(integer_part[0]=='+') {
                     this->_positive = true;
-                    first_index++;
-                } else if (number.at(first_index) == '-') {
+                    integer_part = integer_part.substr(1);
+                }
+                if(integer_part[0]=='-') {
```
Comment (Laouen) [inline include/real/real_explicit.hpp]: should be an "else if" to avoid irrelevant double checks

File: include/real/real_explicit.hpp [hunk 3]
```diff
@@ -221,102 +222,82 @@ namespace boost {
...
+                int exponent = integer_part.length() + add_exponent;
+                if(decimal_part.empty()) {
+                    i = integer_part.length()-1;
+                    while (integer_part[i]=='0'&&i>0)
+                        --i;
+                    integer_part = integer_part.substr(0, i+1);
+                }
+                if(integer_part.empty()) {
+                    i = 0;
+                    while (decimal_part[i]=='0'&&i<decimal_part.length()) {
```
Comment (Laouen) [inline include/real/real_explicit.hpp]: decimal_part[i] == '0' && i < decimal_part.length()
Put spaces between operators and operands.

File: include/real/real_explicit.hpp [hunk 4]
```diff
@@ -221,102 +222,82 @@ namespace boost {
...
+                        ++i;
+                        --exponent;
                     }
+                    decimal_part = decimal_part.substr(i);
                 }
-
-                // If the number is composed by all zeros, then it is zero.
-                if (last_index == first_index && number.at(last_index) == '0') {
+                if(integer_part.empty()&&decimal_part.empty()) {
```
Comment (Laouen) [inline include/real/real_explicit.hpp]: put spaces between operators and operands (this counts everywhere else)

File: include/real/real_explicit.hpp [hunk 5]
```diff
@@ -221,102 +222,82 @@ namespace boost {
...
+                i = decimal_part.length()-1;
+                while (decimal_part[i]=='0'&&i>0) {
+                    --i;
+                }
+                decimal_part = decimal_part.substr(0, i+1);
+                //decimal and integer parts stripped of zeroes
+                int exponent = integer_part.length() + add_exponent;
+                if(decimal_part.empty()) {
+                    i = integer_part.length()-1;
+                    while (integer_part[i]=='0'&&i>0)
```
Comment (Laouen) [inline include/real/real_explicit.hpp]: When you strip zeros from the right side of the integer part, the exponent should be increased. I think this is a bug

File: test/real_eq_operator_test.cpp [hunk 6]
```diff
@@ -940,4 +940,54 @@ TEST_CASE("Operator ==") {
             }
         }
     }
+
+    SECTION("Testing UDL") {
+        SECTION("explicit == explicit") {
```
Comment (Laouen) [inline test/real_eq_operator_test.cpp:945]: Add a test for 12300.00_r (I think it fails due the bug)

**Review by Laouen — CHANGES_REQUESTED** [review]
A few minor code style comments and a possible bug

File: include/real/real_explicit.hpp (hunk 5 above)
Comment (SagnikDey92) [reply include/real/real_explicit.hpp]: I had initially done that as it seemed to be a bug to me as well. But then I noticed that the exponent is already calculated as integer_part.length(). So ++exponent not needed. Plus, if only Travis CI was still working, I could demonstrate that all the tests are passing. But for now, please just take my word for it :-P

File: include/real/real_explicit.hpp (hunk 1 above)
Comment (SagnikDey92) [reply include/real/real_explicit.hpp]: By mistake. I'll add it back.

File: include/real/real_explicit.hpp (hunk 5 above)
Comment (Laouen) [reply include/real/real_explicit.hpp]: 1) In the Test there is no test of the case I say 12300.00_r that is why your tests passes. If you add the test I suggested I'm sure it wouldn't pass.

2) The exponent calculation integer_part.length() is done in the line 273 and you strip zeros in line 276, then if you have 12300.00 first at line 273 it will calculate an exponent of 5 (integer_part = 12300) but later in line 276 you strip the two zeros from the right and the integer_part become 123 while the exponent remains 5 as you are not updating it.

You can fix the bug just by moving the exponent calculation some lines so it is calculated once the integer_part is fully procesed. 

By the way, you could strip zeros using a regex instead of iterating to be consistent in the implementation.

**Review by Laouen — CHANGES_REQUESTED** [review]

File: test/real_eq_operator_test.cpp [hunk 7]
```diff
@@ -988,6 +988,21 @@ TEST_CASE("Operator ==") {
...
+            SECTION("trailing zeroes == without trailing zeroes") {
+
+                SECTION("Without exponent") {
+                    auto a = "12300.00"_r;
+                    auto b = 12300_r;
+
+                    CHECK(a == b);
+                }
+                SECTION("With exponent") {
+                    auto a = "12300.00e-2"_r;
```
Comment (Laouen) [inline test/real_eq_operator_test.cpp]:
```suggestion
                    auto a = "1230000.00e-2"_r;
```

File: test/real_eq_operator_test.cpp [hunk 8]
```diff
@@ -988,6 +988,21 @@ TEST_CASE("Operator ==") {
...
+                    auto b = 123_r;
```
Comment (Laouen) [inline test/real_eq_operator_test.cpp]:
```suggestion
                    auto b = 12300_r;
```

**Review by Laouen — CHANGES_REQUESTED** [review]
